### PR TITLE
impl(docfx): create scaffold for docfx

### PR DESCRIPTION
**Pull request by coryan**
We will need to convert the Doxygen XML output into DocFX to support reference documents at `cloud.google.com`. This just creates the basic structure to host this tool.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10758)

### CODE REVIEW

File: docfx/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Doxygen XML to DocFX for google-cloud-cpp
+
+This directory contains a tool to convert Doxygen XML into DocFX YAML files.
+This tool is used to generate the `google-cloud-cpp` reference documentation
+when hosted at cloud.google.com.
+
+This is not a general-purpose tool. It is not expected to work for other C++
+projects. This tool is only used as part of the documentation pipeline for
+`google-cloud-cpp`. As such, it is only tested in a limited number of platforms.
```
Comment (devbww) [inline docfx/README.md:9]: s/in/on/

File: cmake/GoogleCloudCppFeatures.cmake [hunk 2]
```diff
@@ -219,6 +219,10 @@ endfunction ()
 # additional samples that are enabled if needed.
 function (google_cloud_cpp_enable_features)
     foreach (feature ${GOOGLE_CLOUD_CPP_ENABLE})
+        if ("${feature}" STREQUAL "docfx")
+            add_subdirectory(docfx)
+            continue()
```
Comment (devbww) [inline cmake/GoogleCloudCppFeatures.cmake:224]: If we're going to use `elseif` for all the other cases, we might as well do the same here.  (Or vice versa.)

File: docfx/doxygen2docfx.cc [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <iostream>
+#include <stdexcept>
+
+int main(int argc, char* argv[]) try {
+  auto config = ParseArguments({argv, argv + argc});
+  return 0;
+} catch (std::exception const& ex) {
+  std::cerr << "Standard exception thrown: " << ex.what() << "\n";
+  return 1;
+}
```
Comment (devbww) [inline docfx/doxygen2docfx.cc:25]: No newline at end of file.

File: docfx/parse_arguments_test.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  auto const arguments = std::vector<std::string>({"cmd", "--help"});
+  EXPECT_EXIT(
+      ParseArguments(arguments), [](int status) { return status == 0; },
+      // The usage message goes to stdout, the error log should be empty:
+      "");
+}
+
+TEST(ParseArguments, NoArguments) {
+  EXPECT_THROW(
+      try { ParseArguments({"cmd"}); } catch (std::exception const& ex) {
```
Comment (devbww) [inline docfx/parse_arguments_test.cc:38]: Just an observation, but that formatting seems like it must be a bug.

File: docfx/parse_arguments_test.cc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        EXPECT_EQ(ex.what(), std::string_view("Usage: cmd <infile> <library>"));
```
Comment (devbww) [inline docfx/parse_arguments_test.cc:39]: Perhaps `EXPECT_STREQ(ex.what(), "Usage: cmd <infile> <library>")` is more direct than having to resort to `string_view`.

File: docfx/parse_arguments.cc [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+}  // namespace
+
+Config ParseArguments(std::vector<std::string> const& args) {
+  if (args.size() == 2 && args[1] == "--help") {
+    std::cout << Usage(args[0]) << "\n";
+    std::exit(0);
+  }
+  if (args.size() != 3) throw std::runtime_error(Usage(args[0]));
```
Comment (devbww) [inline docfx/parse_arguments.cc:36]: Nit: No one checks `args[0]` is valid.

Comment (coryan) [reply docfx/parse_arguments.cc:37]: Fixed.

File: docfx/parse_arguments_test.cc (hunk 5 above)
Comment (coryan) [reply docfx/parse_arguments_test.cc]: Fixed here and elsewhere in the file.

File: docfx/parse_arguments_test.cc (hunk 4 above)
Comment (coryan) [reply docfx/parse_arguments_test.cc:48]: Ack

File: docfx/doxygen2docfx.cc (hunk 3 above)
Comment (coryan) [reply docfx/doxygen2docfx.cc]: Fixed

File: cmake/GoogleCloudCppFeatures.cmake (hunk 2 above)
Comment (coryan) [reply cmake/GoogleCloudCppFeatures.cmake]: Fixed

File: docfx/README.md (hunk 1 above)
Comment (coryan) [reply docfx/README.md]: Fixed

File: docfx/parse_arguments.cc [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+std::string Usage(std::string const& cmd) {
+  std::ostringstream os;
+  os << "Usage: " << cmd << " <infile> <library>";
+  return std::move(os).str();
+}
+
+}  // namespace
+
+Config ParseArguments(std::vector<std::string> const& args) {
+  if (args.empty()) throw std::runtime_error(Usage("program-name-missing"));
```
Comment (devbww) [inline docfx/parse_arguments.cc:32]: Aside: Perhaps `ParseArguments()` should take a default program name that could be used here.

File: docfx/parse_arguments_test.cc [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "docfx/parse_arguments.h"
+#include <gmock/gmock.h>
+#include <stdexcept>
+#include <string_view>
```
Comment (devbww) [inline docfx/parse_arguments_test.cc:18]: Obsolete.

**Review by devbww — APPROVED** [review]

File: docfx/parse_arguments.cc (hunk 7 above)
Comment (coryan) [reply docfx/parse_arguments.cc:32]: Ack.  Methinks this does not matter that much.  The only caller outside tests is `main()`.

File: docfx/parse_arguments_test.cc (hunk 8 above)
Comment (coryan) [reply docfx/parse_arguments_test.cc]: Removed.

**Pull request merged**